src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/DeterministicShuffleShardSyncLeaderDecider.java

```diff
@@ -0,0 +1,184 @@
+     * This constructor is package-private to ensure external users use the public
+     * constructor and are thereby aware of the number of periodic shard sync workers
+     * they want in their application.
+     *
+     * @param config       KinesisClientLibConfiguration instance
+     * @param leaseManager LeaseManager instance.
+     */
+    DeterministicShuffleShardSyncLeaderDecider(KinesisClientLibConfiguration config,
+            ILeaseManager<KinesisClientLease> leaseManager) {
+        this(config, leaseManager, PERIODIC_SHARD_SYNC_MAX_WORKERS_DEFAULT);
+    }
+
+    /**
+     * Create an instance overriding the default periodic shard sync worker count.
+     * This is public for use outside KCL.
+     *
+     * @param config                      KinesisClientLibConfiguration instance
+     * @param leaseManager                LeaseManager instance.
+     * @param numPeriodicShardSyncWorkers Max number of periodic shard sync workers.
+     */
+    public DeterministicShuffleShardSyncLeaderDecider(KinesisClientLibConfiguration config,
+            ILeaseManager<KinesisClientLease> leaseManager, int numPeriodicShardSyncWorkers) {
+        this(config, leaseManager, Executors.newScheduledThreadPool(LEADER_DECIDER_THREAD_COUNT),
+                numPeriodicShardSyncWorkers);
+    }
+
+    DeterministicShuffleShardSyncLeaderDecider(KinesisClientLibConfiguration config,
+            ILeaseManager<KinesisClientLease> leaseManager, ScheduledExecutorService leaderElectionThreadPool,
+            int numPeriodicShardSyncWorkers) {
+        this.config = config;
+        this.leaseManager = leaseManager;
+        this.leaderElectionThreadPool = leaderElectionThreadPool;
+        this.numPeriodicShardSyncWorkers = numPeriodicShardSyncWorkers;
+    }
+
+    /*
+     * Shuffles the leases deterministically and elects numPeriodicShardSyncWorkers number of workers
+     * as leaders (workers that will perform shard sync).
+     */
+    private void electLeaders() {
+        try {
+            LOG.debug("Started leader election: " + System.currentTimeMillis());
+            List<KinesisClientLease> leases = leaseManager.listLeases();
+            List<String> uniqueHosts = leases.stream().map(KinesisClientLease::getLeaseOwner)
+                    .filter(owner -> owner != null).distinct().sorted().collect(Collectors.toList());
+
+            Collections.shuffle(uniqueHosts, new Random(DETERMINISTIC_SHUFFLE_SEED));
+            int numShardSyncWorkers = Math.min(uniqueHosts.size(), numPeriodicShardSyncWorkers);
+            // In case value is currently being read, we wait for reading to complete before updating the variable.
+            // This is to prevent any ConcurrentModificationException exceptions.
+            readWriteLock.writeLock().lock();
+            leaders = new HashSet<>(uniqueHosts.subList(0, numShardSyncWorkers));
+            LOG.debug("Elected leaders: " + String.join(", ", leaders));
```

Comment:
Please use the placeholder {} while logging as we have used elsewhere in KCL.

Reply:
Discussed offline.